src/hotspot/share/oops/methodData.cpp

```diff
@@ -1201,6 +1202,15 @@ void MethodData::post_initialize(BytecodeStream* stream) {
   }
 }
 
+// Special ctor for ciMethodData.
+MethodData::MethodData(ciMethodData& data)
+  : _extra_data_lock(Mutex::leaf, "unused") {
+  _extra_data_lock.~Mutex(); // release allocated resources before zeroing
```

Comment:
So if _extra_data_lock was a pointer to Mutex, then this statement would be simply "delete _extra_data_lock;" instead, but we'd still have this copy constructor, right?

Reply:
Why isn't this a regular copy constructor taking MethodData, and pass this._orig instead and leave the zeroing in the caller?  I don't like that oops/methodData knows about ciMethodData.

Reply:
If `_extra_data_lock` were a pointer, then `MethodData::MethodData` would just `_extra_data_lock(NULL)`.

Regarding where to put zeroing, I'm fine doing it either way. 

Just want to give a try the idea David and Kim proposed with separating the lock from MDO (w/ superclass trick).